smartapps/curb/curb-energy-monitor.src/curb-energy-monitor.groovy

```diff
@@ -0,0 +1,620 @@
+            atomicState.authToken = resp.data.access_token
+
+            getCurbLocations()
+        }
+        if (atomicState.authToken) {
+            success()
+        } else {
+            fail()
+        }
+    } else {
+        log.error "callback() failed oauthState != atomicState.oauthInitState"
+    }
+}
+
+private removeChildDevices(delete) {
+    delete.each {
+        deleteChildDevice(it.deviceNetworkId)
+    }
+}
+
+def getCurbLocations() {
+    //log.debug("Requesting Curb location info");
+
+    def params = [
+        uri: "http://app.energycurb.com",
+        path: "/api/locations",
+        headers: ["Authorization": "Bearer ${atomicState.authToken}"]
+    ]
+
+    try {
+        httpGet(params) {
+            resp ->
+                def locationNameList = []
+            def locationLookup = []
+            //log.debug(resp.data)
+            resp.data.each {
+                locationNameList.push(it.name)
+                locationLookup.push(it)
+            }
+            atomicState.locationNames = locationNameList
+            atomicState.locationLookup = locationLookup
+
+            //log.debug("Location ID: ${atomicState.location}")
+        }
+    } catch (e) {
+        log.error "something went wrong: $e"
+    }
+}
+
+def updateSelectedLocationId() {
+    atomicState.locationLookup.each {
+        // log.debug "location match check: ${it.name} vs selected: ${settings.curbLocation}"
+        if (it.name == settings.curbLocation) {
+            atomicState.location = it.id
+            return it.id
+        }
+    }
+}
+
+def updateChildDevice(dni, label, values) {
+    try {
+        def existingDevice = getChildDevice(dni)
+        existingDevice.handleMeasurements(values)
+    } catch (e) {
+        log.error "Error creating or updating device: ${e}"
+    }
+}
+
+def createChildDevice(dni, label) {
+    return addChildDevice("curb", "Curb Power Meter", dni, null, [name: "${dni}", label: "${label}"])
+}
+
+def getDevices() {
+    def params = [
+        uri: "https://app.energycurb.com",
+        path: "/api/latest/${atomicState.location}",
+        headers: ["Authorization": "Bearer ${atomicState.authToken}"],
+        requestContentType: 'application/json'
+    ]
+    asynchttp_v1.get(processDevices, params)
+}
+
+def getAllData() {
+
+    // def aggregateResolution = "h"
+    // if (settings.energyInterval == "30m" || settings.energyInterval == "15m" || settings.energyInterval == "1m") {
+    //     aggregateResolution = "s"
+    // }
+    // if (settings.energyInterval == "1h") {
+    //     aggregateResolution = "m"
+    // }
+
+    def onehparams = [
+        uri: "https://app.energycurb.com",
+        path: "/api/aggregate/${atomicState.location}/1h/m",
+        headers: ["Authorization": "Bearer ${atomicState.authToken}"],
+        requestContentType: 'application/json'
+    ]
+    log.debug onehparams
+    asynchttp_v1.get(processKwh1h, onehparams)
```

Comment:
There are 5 asynchHttp_v1 calls that will result in 5 different callbacks, are the responses spread out in time or is there a possibility that responses arrive more or less at the same time?
Reason I'm asking is that it is preferable to make one request and handled one response, especially since this is scheduled to be done once every minute and asynchttp is used. Each of the response will cause an instant of the app to be spun up which will spin up an instance of the child for each of the call that app instance makes, keep in mind that the state in the DTH is not thread safe.
Granted, if the response for the http get takes to long time it will cause a socket timeout, but this is scheduled to be done every minute so the response should not take that long time...